### PR TITLE
[srvls][SRVOCF-278] Adding Python functions docs

### PR DESCRIPTION
**Pull request by abrennan89**
Applies for OCP 4.6+
When merging, functions docs must be commented out of topicmap until TP release (end of May)

### CODE REVIEW

**Comment by abrennan89** [comment]
Python docs QE preview link: https://deploy-preview-32181--osdocs.netlify.app/openshift-enterprise/latest/serverless/functions/serverless-developing-python-functions.html

File: modules/serverless-python-function-return-values.adoc [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+Functions may return any value supported by https://flask.palletsprojects.com/en/1.1.x/quickstart/#about-responses[Flask] because the invocation framework proxies these values directly to the Flask server.
+
+.Example
+[source,python]
+----
+def main(context: Context):
+    data = { "message": "Howdy!" }
+    headers = { "content-type": "application/json" }
+    return body, 200, headers
```
Comment (jrangelramos) [inline modules/serverless-python-function-return-values.adoc:16]: I'm not a much familiar with python and Flask api, but I think this will not work since `body` was not defined. 
I tried that I got an error. 

So maybe it would be good to change to:

`body = { "message": "Howdy!" }`
or
`return data, 200, headers`

File: modules/serverless-python-template.adoc [hunk 2]
```diff
@@ -0,0 +1,26 @@
+// Module included in the following assemblies
+//
+// * /serverless/functions/serverless-developing-python-functions.adoc
+
+[id="serverless-python-template_{context}"]
+= Python function template structure
+
```
Comment (jrangelramos) [inline modules/serverless-python-template.adoc:7]: Just suggestion: one thing I believe is nice to have in the doc is the command to create a python function (`kn func create -l python`).

File: modules/serverless-invoking-python-functions.adoc [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+.Example context object
+[source,python]
+----
+def main(context: Context):
+    """
+    The context parameter contains the Flask request object and any
+    CloudEvent received with the request.
+    """
+    print(f"Method: {context.request.method}")
+    print(f"Event data {context.cloud_event.data})
```
Comment (jrangelramos) [inline modules/serverless-invoking-python-functions.adoc:28]: Here I see it need to close quotes.

`print(f"Event data {context.cloud_event.data}")`

File: modules/serverless-python-function-return-values.adoc (hunk 1 above)
Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc:16]: @lance wdyt?

File: modules/serverless-python-template.adoc [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+.Template structure
+[source,terminal]
+----
+fn
+├── func.py <1>
+├── func.yaml <2>
+├── requirements.txt <3>
+└── test_func.py <4>
+----
+<1> Python functions have very few restrictions. The only real requirement is that your project contain a `func.py` file which contains a `main()` function.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-python-template.adoc:19]: Good point, let me modify this a bit

File: modules/serverless-python-template.adoc (hunk 2 above)
Comment (abrennan89) [reply modules/serverless-python-template.adoc:7]: I will be adding the `kn func create` docs at another time, they need their own module 😉 this is just about what's contained in a Python function, not the `kn` tasks for it really.

File: modules/serverless-python-function-return-values.adoc (hunk 1 above)
Comment (lance) [reply modules/serverless-python-function-return-values.adoc:16]: Oh yeah - good catch. Setting `body` is probably the more illustrative approach. `body = { "message": "Howdy!" }`

Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc:16]: Adding in the next commit, lmk if it looks correct 🙂

**Comment by abrennan89** [comment]
@jrangelramos made some changes in this commit: https://github.com/openshift/openshift-docs/pull/32181/commits/e6e297e1cdc423e52181b9ef03c44e32cc8b20e1
The preview link should update in a few minutes too.

Thanks for the feedback!

File: modules/serverless-python-function-return-values.adoc [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+.Example
+[source,python]
+----
+def main(context: Context):
+    data = { "message": "Howdy!" }
+    headers = { "content-type": "application/json" }
+    body = { "message": "Howdy!" }
+----
+
```
Comment (jrangelramos) [inline modules/serverless-python-function-return-values.adoc:18]: I think here we could have just body and keep the return sentence (something like that, was I meant first time):

```
def main(context: Context):
    body = { "message": "Howdy!" }
    headers = { "content-type": "application/json" }
    return body, 200, headers
```

**Comment by jrangelramos** [comment]
QE Approved :)

File: modules/serverless-python-template.adoc [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+├── func.yaml <2>
+├── requirements.txt <3>
+└── test_func.py <4>
+----
+<1> Contains a `main()` function.
+<2> Used to determine the image name and registry.
+<3> Additional dependencies can be added to the `requirements.txt` file as they would be in any other Python project.
+<4> Contains a simple unit test that can be used to test your function locally.
+
+== Dependencies
```
Comment (kalexand-rh) [inline modules/serverless-python-template.adoc:26]: Please double-check the rendered assembly to make sure that using this short section is doing what you want it to do. (I might remove the separate heading or turn it into a note.)

File: serverless/functions/serverless-developing-python-functions.adoc [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+toc::[]
+
+:FeatureName: {FunctionsProductName}
+include::modules/technology-preview.adoc[leveloffset=+2]
+
+Before you can develop functions, you must complete the set up steps in xref:../../serverless/functions/serverless-functions-setup.adoc#serverless-functions-setup[Setting up {FunctionsProductShortName}].
+
+include::modules/serverless-python-template.adoc[leveloffset=+1]
+include::modules/serverless-invoking-python-functions.adoc[leveloffset=+1]
```
Comment (kalexand-rh) [inline serverless/functions/serverless-developing-python-functions.adoc:15]: I feel like I'm missing part of the story of this assembly. The modules about invoking and return values seem like they're part of a bigger process.

**Review by kalexand-rh — COMMENTED** [review]
I have a few style picks and suggestions. I think this new assembly's shaping up well!

File: serverless/functions/serverless-developing-python-functions.adoc [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+= Developing Python functions
+:context: serverless-developing-python-functions
+include::modules/common-attributes.adoc[]
+
+toc::[]
+
+:FeatureName: {FunctionsProductName}
+include::modules/technology-preview.adoc[leveloffset=+2]
+
+Before you can develop functions, you must complete the set up steps in xref:../../serverless/functions/serverless-functions-setup.adoc#serverless-functions-setup[Setting up {FunctionsProductShortName}].
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply serverless/functions/serverless-developing-python-functions.adoc:12]: Added for both in next commit, thanks!

File: modules/serverless-testing-python-functions.adoc [hunk 9]
```diff
@@ -0,0 +1,29 @@
+// Module included in the following assemblies
+//
+// * /serverless/functions/serverless-developing-python-functions.adoc
+
+[id="serverless-testing-python-functions_{context}"]
+= Testing Python functions
+
+Python functions can be tested locally on your computer. In the default project that is created when you create a function using `kn func create`, there is a `test_func.py` file which contains a simple unit test.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-testing-python-functions.adoc:8]: Did some rewriting of this part in the next commit

File: modules/serverless-python-function-return-values.adoc [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+----
+@event("event_source"="/my/function", "event_type"="my.type")
+def main(context):
+    # business logic here
+    data = do_something()
+    # more data processing
+    return data
+----
+
+This example sends a CloudEvent as the response value, with a type of `"my.type"`, a source of `"/my/function"`, and the data property set to `data`. The `event_source` and `event_type` decorator attributes are both optional.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc:36]: @lance on this one does it mean to call out

1. "The `return` property is set to `data`"
2. "The `data` property is set to `do_something()`"
3. something else?

This originally came from the content [here](https://github.com/boson-project/func/blob/main/docs/guides/python.md#cloudevents) so I'm not positive which you meant.

File: modules/serverless-python-function-return-values.adoc [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+If not specified, the CloudEvent `event_source` attribute is set to `"/parliament/function"`, and the `event_type` attribute is set to `"parliament.response"`.
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc:38]: @lance can you provide explanations for the `parliament` ones cos I'm not sure what they do and it's not explained [here](https://github.com/boson-project/func/blob/main/docs/guides/python.md#cloudevents) - thanks!

File: serverless/functions/serverless-functions-getting-started.adoc [hunk 12]
```diff
@@ -12,7 +12,7 @@ include::modules/technology-preview.adoc[leveloffset=+2]
 This guide explains how you can create and manage a function on an {ServerlessProductName} installation by using the `kn` CLI.
 // TODO: add info about developer console at 4.8 when this is supported.
 
-[id="prerequisites_serverless-functions-getting-started]
+[id="prerequisites_serverless-functions-getting-started"]
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply serverless/functions/serverless-functions-getting-started.adoc:15]: Hopefully better now?

File: modules/serverless-python-function-return-values.adoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (lance) [reply modules/serverless-python-function-return-values.adoc:36]: @abrennan89 the `data` variable returned from the `main()` function is sent as the `data` property on the CloudEvent response. How about rewording it similar to this, with a link to the CloudEvent spec? Does that help?

```suggestion
This example sends a CloudEvent as the response value, with a type of `"my.type"` and a source of `"/my/function"`. The returned `data` variable is set as the CloudEvent's [`data` property](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#event-data). The `event_source` and `event_type` decorator attributes are both optional.
```

File: modules/serverless-python-function-return-values.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (lance) [reply modules/serverless-python-function-return-values.adoc:38]: @abrennan89 the `event_source` and `event_type` attributes refer to properties in a Python CloudEvent, as manifested by [cloudevents/sdk-python](https://github.com/cloudevents/sdk-python). These properties are required per the CloudEvents spec. For convenience, since the user may not care about the type and/or source values, we populate them with default values. It's hard to get into a lot of detail on all of this because we are simply referencing external CNCF libraries and specifications.

File: modules/serverless-invoking-python-functions.adoc [hunk 13]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies
+//
+// * /serverless/functions/serverless-developing-python-functions.adoc
+
+[id="serverless-invoking-python-functions_{context}"]
+= Invoking Python functions
```
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-invoking-python-functions.adoc:6]: I believe there are other topics in the docs that aren't tasks and start with gerunds, e.g. [Understanding the update service](https://docs.openshift.com/container-platform/4.7/updating/understanding-the-update-service.html). I'm not really sure how else we would title this since it explains how a user would invoke functions?

File: modules/serverless-python-function-return-values.adoc (hunk 10 above)
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc:36]: Updated in next commit, PTAL

File: modules/serverless-python-function-return-values.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (abrennan89) [reply modules/serverless-python-function-return-values.adoc]: Above comment removes mention of them anyway, probably no need to mention if it's just a default setting and we're linking to the CloudEvent spec

File: modules/serverless-invoking-python-functions.adoc (hunk 13 above)
[the remark this answers is not in the document]
Comment (kalexand-rh) [reply modules/serverless-invoking-python-functions.adoc:6]: They shouldn't, though. I think those have slid in because "Understanding" makes it pretty clear that it's conceptual, where "Invoking" doesn't. (I might open a PR to s/Understanding/About someday.)

In the long run, if possible, it might be best to change this into a proper procedure. In the short term, I might call it something that makes it clearer that it's more conceptual than procedural, like:
- "Python function invocation"
- "About invoking Python functions"
- "Sample that invokes a Python function"
